### ISSUE TITLE
long lived object in tpool.Proxy

### ISSUE THREAD

**Issue by boytm** [eventlet/eventlet#469]
See the following code,  after test_batch_exception() there are 20 (tpool thread count) A instance \_\_del\_\_ until program exit, even the A instance's wrap object tpool.Proxy is \_\_del\_\_ .

```python

import sys
import os
from eventlet import tpool, sleep
import gc
import objgraph
import traceback

class A(object):
    def __init__(self, name):
        self.name = name
    
    def __str__(self):
        return 'A(%s)' % self.name
    
    def write(self, s):
        print('write %s' % s)
        
    def raise_error(self):
        raise IOError('not implement yet')
    
    def __del__(self):
        print('__del__ %s' % self.name)
        
def test_batch_exception(count):
    for i in range(count):
        try:
            tpool.Proxy(A(str(i))).raise_error()
        except IOError:
            pass
        
def test_simple():     

    def proxy_del(x):
        sys.stdout.write('proxy _del__ %s contain %s\n' % (str(x), str(x._obj)))
    tpool.Proxy.__del__ = proxy_del
    
    a=A('aaa')
    b=A('bbb')
    pa = tpool.Proxy(a)
    pb = tpool.Proxy(b)
    
    pb.write('some thing...')
    try:
        pa.raise_error()
    except IOError as e:
        print(e)

    pa = None
    pb = None
    b = None
    gc.collect()
    objgraph.show_backrefs(a, max_depth=500, filename = "D:\\a.dot")
    
    sys.exc_clear()   # clear sys.exc_traceback reference
    gc.collect()
    print('after exc_clear(), gc')
    objgraph.show_backrefs(a, max_depth=500, filename = "D:\\b.dot")
    
    sleep(0)          # switch to tpool_trampoline(), in order to release result rv's reference
    gc.collect()
    print('after switch eventlet, gc')
    objgraph.show_backrefs(a, max_depth=500, filename = "D:\\c.dot")
    
    a = None
    
    print 'end simple'

def main():
    test_simple()
    print 'after simple'
    #sys.exc_clear()
    #sys.exc_info()
    print 'after exc_info()'
    gc.collect()
    print 'after gc'
    print 'end main'

#main()
test_batch_exception(100)  # remain object equal tpool thread count
gc.collect()
print '-----'
```
If you want to see who have a reference to object A, uncomment main(), then objgraph will dump the detail.  

In short, every thread in tpool's thread pool have a reference to current function ( tworker() ) exception‘s backtrace. The backtrace hold a reference to function where exception occurred. And is cleared until function exit, so we must clear it manually by sys.exc_clear() 

https://github.com/boytm/eventlet/commit/9ccf6d136e60928e9ea69f2e8559be38298a0184

**Comment by temoto** [comment]
@boytm Just for information, turns out python3 doesn't have `sys.exc_clear()`, they say `except` statement emits semantically same bytecode.

**Comment by temoto** [comment]
@boytm this is final version of patch, seems to over-satisfy all recommendations on exception handling.

```
        try:
            rv = meth(*args, **kwargs)
        except SYS_EXCS:
            raise
        except EXC_CLASSES:
            rv = sys.exc_info()
            if six.PY3:
                traceback.clear_frames(rv[1].__traceback__)
        if six.PY2:
            sys.exc_clear()
```

**Comment by temoto** [comment]
Fix is merged 69983c6d76e7c122d14a804b5932dfef00d8f7c8

Thank you. Have a great time.

**Comment by boytm** [comment]
@temoto 
```python     
        try:
            rv = meth(*args, **kwargs)
        except SYS_EXCS:
            raise
        except EXC_CLASSES:
            rv = sys.exc_info()
            if sys.version_info >= (3, 4):         # python 3, no need
                traceback.clear_frames(rv[1].__traceback__)
        if six.PY2:                                          # when no exception, no need 
            sys.exc_clear()
```

In python 3, the sys.exc_info() is only valid in "except: " clause or function called by "except: " clause. see https://docs.python.org/3/library/sys.html#sys.last_traceback

>This function returns a tuple of three values that give information about the exception that is currently being handled. The information returned is specific both to the current thread and to the current stack frame. If the current stack frame is not handling an exception, the information is taken from the calling stack frame, or its caller, and so on until a stack frame is found that is handling an exception. Here, “handling an exception” is defined as “executing an except clause.” For any stack frame, only information about the exception being currently handled is accessible.

You can test the following code, python 27 can access the sys.exc_info() until test_simple() exit, but not possible in python 35 and python 36

```python
import sys

def call_in_except():
    print('a call in except: %s', str(sys.exc_info()))
        
def test_simple():     
    pa = []
    try:
        pa.raise_error()
    except AttributeError as e:
        call_in_except()
        print('in except: %s', str(sys.exc_info()))
    finally:
        print('in finally: %s', str(sys.exc_info()))

    print('in same function: %s', str(sys.exc_info()))

test_simple()
print('out function: %s', str(sys.exc_info()))

```
